MigrationGuide.md

```diff
@@ -0,0 +1,224 @@
+     }
+     else
+     {
+         // Other application functions.
+     }
+}
+```
+**New Code Snippet**:
+```
+// Variables used in this example.
+MQTTStatus_t status;
+// This context is assumed to be initialized and connected.
+MQTTContext_t * pContext;
+
+while( true )
+{
+     status = MQTT_ProcessLoop( pContext );
+
+     if( status != MQTTSuccess && status != MQTTNeedMoreBytes )
+     {
+         // Determine the error. It's possible we might need to disconnect
+         // the underlying transport connection.
+     }
+     else
+     {
+         // Other application functions.
+     }
+}
+```
+
+The `MQTT_ReceiveLoop` function no longer uses a timeout as this led to unavoidable busy-waiting. Thus, the signature of `MQTT_ReceiveLoop` changed from `MQTTStatus_t MQTT_ReceiveLoop( MQTTContext_t * pContext, uint32_t timeoutMs )`  to  `MQTTStatus_t MQTT_ReceiveLoop( MQTTContext_t * pContext )`. Additionally, `MQTT_ReceiveLoop` can now return `MQTTNeedMoreBytes`. A return of `MQTTNeedMoreBytes` means that `MQTT_ReceiveLoop` received only a part of an MQTT packet and will need to be called again (probably after a bit of delay) in order to finish receiving the MQTT packet. Thus, to migrate, simply remove the timeout from the `MQTT_ReceiveLoop` call and additionally check for if `MQTTNeedMoreBytes` has been returned when checking the status of `MQTT_ReceiveLoop`. For example:
+
+**Old Code Snippet**:
+```
+// Variables used in this example.
+MQTTStatus_t status;
+uint32_t timeoutMs = 100;
+// This context is assumed to be initialized and connected.
+MQTTContext_t * pContext;
+
+while( true )
+{
+     status = MQTT_ReceiveLoop( pContext, timeoutMs );
+
+     if( status != MQTTSuccess )
+     {
+         // Determine the error. It's possible we might need to disconnect
+         // the underlying transport connection.
+     }
+     else
+     {
+         // Other application functions.
+     }
+}
+```
+**New Code Snippet**:
+```
+// Variables used in this example.
+MQTTStatus_t status;
+// This context is assumed to be initialized and connected.
+MQTTContext_t * pContext;
+
+while( true )
+{
+     status = MQTT_ReceiveLoop( pContext );
+
+     if( status != MQTTSuccess && status != MQTTNeedMoreBytes )
+     {
+         // Determine the error. It's possible we might need to disconnect
+         // the underlying transport connection.
+     }
+     else
+     {
+         // Other application functions.
+     }
+}
+```
+
+The `TransportInterface_t` structure now has a new member `writev` that can be defined to allow sending multiple MQTT message vectors as a single packet. This can improve performance, but is not necessary to implement. However, when programs initialize a `TransportInterface_t` structure, they **MUST** either set `writev` to a working implementation or set it `NULL`. Not doing this will lead to undefined behavior as the coreMQTT library checks if `writev` is `NULL` to determine if it should be used. For example:
```

Comment:
```suggestion
The `TransportInterface_t` structure now has a new member `writev`. It uses scatter-gather approach to send multiple MQTT packet components as a single packet to reduce overhead and improve performance. However, it is COMPLETELY OPTIONAL to implement. To that end, when application(s) initialize a `TransportInterface_t` structure, they **MUST** either set `writev` to a working implementation or set it `NULL`. Not doing this will lead to undefined behavior as the coreMQTT library checks if `writev` is `NULL` to determine if it should be used. For example:
```